### PR TITLE
[rollout, trainer] fix: handle empty token sequences in _pad_token_ids and metric timing

### PR DESCRIPTION
**Pull request by GJWu-zyx**
### What does this PR do?

Fix two crashes triggered by empty token sequences when `async_training.partial_rollout=False` and parameter synchronization aborts in-flight vLLM requests during prefill:

  1. **`_pad_token_ids` in agent_loop.py** — `tokenizer.pad()` with empty input returns a dict with list values instead of tensors, even when `return_tensors="pt"` is set, causing `AttributeError:
  'list' object has no attribute 'dim'`.
  2. **`compute_timing_metrics` in metric_utils.py** — An empty response batch yields zero token counts, causing `ZeroDivisionError`.

  Both are valid edge cases: `partial_rollout=False` is a supported configuration, and the code should handle aborted requests that return no generated tokens.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

  - Tested with `partial_rollout=False` + vLLM dump enabled (slows inference, increases abort probability). Without the fix, both crashes reproduce consistently. With the fix, training runs to 
  completion without errors.
  - Metric output shows `timing_per_token_ms/*` values as `0` when zero tokens are processed, instead of crashing.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [x] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/experimental/agent_loop/agent_loop.py [hunk 1]
```diff
@@ -609,6 +609,13 @@ def _pad_token_ids(
         return_attention_mask: bool,
     ) -> dict[str, torch.Tensor]:
         """Right/left pad a flat list of token ids to a ``(1, max_length)`` tensor."""
+        # tokenizer.pad() with empty input returns dict with list values
+        # instead of tensors, which breaks downstream .dim() calls.
+        if not tokens:
+            result = {"input_ids": torch.zeros((1, max_length), dtype=torch.long)}
+            if return_attention_mask:
+                result["attention_mask"] = torch.zeros((1, max_length), dtype=torch.long)
+            return result
```
[the remark this answers is not in the document]
Comment (Luosuu) [reply verl/experimental/agent_loop/agent_loop.py:619]: @GJWu-zyx please fix

File: verl/trainer/ppo/metric_utils.py [hunk 2]
```diff
@@ -304,7 +304,7 @@ def compute_timing_metrics(batch: DataProto, timing_raw: dict[str, float]) -> di
     return {
         **{f"timing_s/{name}": value for name, value in timing_raw.items()},
         **{
-            f"timing_per_token_ms/{name}": timing_raw[name] * 1000 / num_tokens_of_section[name]
+            f"timing_per_token_ms/{name}": timing_raw[name] * 1000 / max(num_tokens_of_section[name], 1)
```
[the remark this answers is not in the document]
Comment (Luosuu) [reply verl/trainer/ppo/metric_utils.py]: @GJWu-zyx please fix

**Review by Luosuu — CHANGES_REQUESTED** [review]

File: verl/experimental/agent_loop/agent_loop.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (GJWu-zyx) [reply verl/experimental/agent_loop/agent_loop.py:619]: @Luosuu Fixed! Both suggestions are now applied — using pad_token_id instead of hardcoded 0, and returning 0.0 for empty token timing metrics. PTAL 🙏

File: verl/trainer/ppo/metric_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (GJWu-zyx) [reply verl/trainer/ppo/metric_utils.py]: @Luosuu Fixed! Both suggestions are now applied — using pad_token_id instead of hardcoded 0, and returning 0.0 for empty token timing metrics. PTAL 🙏

File: verl/experimental/agent_loop/agent_loop.py [hunk 3]
```diff
@@ -609,6 +609,14 @@ def _pad_token_ids(
         return_attention_mask: bool,
     ) -> dict[str, torch.Tensor]:
         """Right/left pad a flat list of token ids to a ``(1, max_length)`` tensor."""
+        # tokenizer.pad() with empty input returns dict with list values
+        # instead of tensors, which breaks downstream .dim() calls.
+        if not tokens:
+            pad_id = self.tokenizer.pad_token_id if self.tokenizer.pad_token_id is not None else 0
```
Comment (Luosuu) [inline verl/experimental/agent_loop/agent_loop.py:615]: Please add a small regression test for this empty-token path, ideally with a tokenizer whose `pad_token_id != 0`, so we lock in that `input_ids` use the tokenizer pad id while `attention_mask` stays all zero.

File: verl/trainer/ppo/metric_utils.py [hunk 4]
```diff
@@ -304,7 +304,9 @@ def compute_timing_metrics(batch: DataProto, timing_raw: dict[str, float]) -> di
     return {
         **{f"timing_s/{name}": value for name, value in timing_raw.items()},
         **{
-            f"timing_per_token_ms/{name}": timing_raw[name] * 1000 / num_tokens_of_section[name]
+            f"timing_per_token_ms/{name}": (
```
Comment (Luosuu) [inline verl/trainer/ppo/metric_utils.py:307]: Please add a zero-token regression test for `compute_timing_metrics`; the existing CPU test only covers nonzero token counts, so this edge case could regress back to divide-by-zero or misleading per-token timing.

File: verl/experimental/agent_loop/agent_loop.py (hunk 3 above)
Comment (GJWu-zyx) [reply verl/experimental/agent_loop/agent_loop.py:615]: @Luosuu Both regression tests added. PTAL 🙏

File: verl/trainer/ppo/metric_utils.py (hunk 4 above)
Comment (GJWu-zyx) [reply verl/trainer/ppo/metric_utils.py:307]: @Luosuu Both regression tests added. PTAL 🙏

**Review by Luosuu — APPROVED** [review]
LGTM, thanks for adding the regression tests. I will merge when CI looks good.
